llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -705,180 +625,256 @@ void EntryPointGroup::rebuildFromNames(const std::vector<std::string> &Names,
-    return Ret;
+  // Creates a rule, which adds a list of dash-separated integers converted
+  // into strings listed in a metadata to a resulting identifier.
+  void registerListOfIntegersInMetadataRule(StringRef MetadataName) {
+    Rules.emplace_back(Rule::RKind::K_IntegersListMetadata, MetadataName);
   }
 
-  static UsedOptionalFeatures getEmpty() {
-    UsedOptionalFeatures Ret;
-    Ret.IsEmpty = true;
-    return Ret;
+  // Creates a rule, which adds a list of sorted dash-separated integers
+  // converted into strings listed in a metadata to a resulting identifier.
+  void registerListOfIntegersInMetadataSortedRule(StringRef MetadataName) {
+    Rules.emplace_back(Rule::RKind::K_SortedIntegersListMetadata, MetadataName);
   }
 
 private:
-  // For DenseMap:
-  llvm::hash_code Hash = {};
-  bool IsTombstoneKey = false;
-  bool IsEmpty = false;
-
-public:
-  bool operator==(const UsedOptionalFeatures &Other) const {
-    // Tombstone does not compare equal to any other item
-    if (IsTombstoneKey || Other.IsTombstoneKey)
-      return false;
+  struct Rule {
+  private:
+    std::variant<StringRef, std::tuple<StringRef, StringRef, StringRef>,
+                 std::function<std::string(Function *)>>
+        Storage;
+
+  public:
+    enum class RKind {
+      K_Callback,
+      K_SimpleStringAttribute,
+      K_FlagMetadata,
+      K_FlagAttribute,
+      K_IntegersListMetadata,
+      K_SortedIntegersListMetadata
+    };
+    RKind Kind;
+
+    // Returns an index into std::variant<...> Storage defined above, which
+    // corresponds to the specified rule Kind.
+    constexpr static std::size_t storage_index(RKind K) {
+      switch (K) {
+      case RKind::K_SimpleStringAttribute:
+      case RKind::K_IntegersListMetadata:
+      case RKind::K_SortedIntegersListMetadata:
+        return 0;
+      case RKind::K_Callback:
+        return 2;
+      case RKind::K_FlagMetadata:
+      case RKind::K_FlagAttribute:
+        return 1;
+      }
+      // can't use llvm_unreachable in constexpr context
+      return std::variant_npos;
+    }
 
-    if (Aspects.size() != Other.Aspects.size())
-      return false;
+    template <RKind K> auto getStorage() const {
+      return std::get<storage_index(K)>(Storage);
+    }
 
-    for (size_t I = 0, E = Aspects.size(); I != E; ++I) {
-      if (Aspects[I] != Other.Aspects[I])
-        return false;
+    template <typename... Args>
+    Rule(RKind K, Args... args) : Storage(args...), Kind(K) {
+      assert(storage_index(K) == Storage.index());
     }
 
-    return IsEmpty == Other.IsEmpty && UsesLargeGRF == Other.UsesLargeGRF;
-  }
+    Rule(Rule &&Other) = default;
+  };
 
-  unsigned hash() const { return static_cast<unsigned>(Hash); }
+  std::vector<Rule> Rules;
 };
 
-struct UsedOptionalFeaturesAsKeyInfo {
-  static inline UsedOptionalFeatures getEmptyKey() {
-    return UsedOptionalFeatures::getEmpty();
-  }
+std::string FunctionsCategorizer::computeCategoryFor(Function *F) const {
```

Comment:
Do we want to add a delimiter in the result string between rules? One corner case is when two attributes have values like this kernel1: 1 and 20; kernel2: 12 and 0. When we concatenate string for both cases we get 120. 

Thanks

Reply:
Good point, thanks! Added a delimiters in 67082dda9a95d3dabaa3c4c9158e86f8b1b5ba27